sql/catalyst/src/main/scala-2.12/org/apache/spark/sql/catalyst/util/GenericArrayDataHelper.scala

```diff
@@ -0,0 +1,28 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.util
+
+object GenericArrayDataHelper {
+  def toArray(seq: Seq[Any]): Array[Any] = seq.toArray
+
+  def toArray(seqOrArray: Any): Array[Any] = seqOrArray match {
+    case seq: Seq[Any] => seq.toArray
+    case array: Array[Any] => array  // array of objects, so no need to convert
```

Comment:
Wait, do you mean AnyRef here? that's the supertype of objects as opposed to primitives. But while all object arrays have the same type, each primitive array is its own type. I think this happens to 'work' as a result -- an object array would match an array of anything, while a primitive array would match only Array[_]. There's probably a cleaner way to do this, but maybe not a faster one; maybe at least say "AnyRef" here. But I guess that's how the current code works